### PR TITLE
fix: add unwind safety to `resource_scope`

### PR DESCRIPTION
**Pull request by joseph-gio**
# Objective

The function `World::resource_scope` is a primitive that allows the user to get mutable access to a resource and the rest of the world at the same time. This is incredibly useful for exclusive systems, implementations of `Command`, and custom abstractions. It works by temporarily removing the resource from the world and running a user-provided closure with mutable access to both. After the closure completes, the resource is re-inserted into the world for other code to use.

With the current implementation, any panics originating in the user-provided closure will result in the resource being lost. This is problematic for a couple of reasons:
* Any users wishing to utilize `std::panic::catch_unwind` have to deal with the fact that the panic has left the world in an invalid state.
  * While some individuals consider catch_unwind to be bad practice, it is a native feature of Rust that we should provide support for. In certain situations it is quite valuable:
    * When creating apps targeting production, it is crucial to avoid hard-crashing on users, and in many cases it is not reasonable to ensure that every code path will always complete successfully. catch_unwind is useful for capturing panics and shifting into a controlled error state. This is hard to do when the world is left in an invalid state.
    * It is often useful to _localize_ panics, rather than fully recover from them. When using threads or async tasks, the usual default is to localize panics to the original context that caused them. When implementing abstractions that give async tasks access to the world, it is useful to capture panics and propagate them back to the async context.
* Panics occurring inside of `resource_scope` can lead to knock-on errors, which can make debugging more difficult, as the user may need to sort through many errors to identify which one was the root cause.

The objective here is to make resource_scope panic-safe, ensuring that the resource is always restored to the world when unwinding from a panic (on platforms that support unwinding; `panic=abort` is unrecoverable by design). This enables more robust panic handling, and allows for more graceful shutdown sequences when panics are truly fatal.

## Solution

Add a "guard" type that captures mutable access to the world and temporary ownership of the resource. This type has a `Drop` implementation that is responsible for re-inserting the resource at the end of the scope -- this pattern is similar to `try ... finally` blocks in many languages.

## Testing

* Added a test that verifies the resource is properly restored when unwinding from `resource_scope`.
* Added a regression test for an existing edge-case triggered by `World::clear_resources` or `World::clear_all`.

### CODE REVIEW

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -2726,40 +2726,114 @@ impl World {
...
+                // take ownership of the value first so it'll get dropped if we return early
+                // SAFETY: drop semantics ensure that `self.value` will never be accessed again after this call
+                let value = unsafe { ManuallyDrop::take(&mut self.value) };
+
+                let Some(resource_data) = self.world.storages.resources.get_mut(self.component_id)
+                else {
+                    return;
+                };
+
+                // in debug mode, raise a panic if user code re-inserted a resource of this type within the scope.
```
Comment (hymm) [inline crates/bevy_ecs/src/world/mod.rs:2758]: Changing this to only panic in debug could be considered a breaking change. We might want to add a migration guide. I could go either way on this as this is a pretty minor change.

File: crates/bevy_ecs/src/world/mod.rs
Comment (hymm) [inline crates/bevy_ecs/src/world/mod.rs:2711]: This error message is potentially incorrect now if we're in debug and `f` panicked.

File: crates/bevy_ecs/src/world/mod.rs [hunk 2]
```diff
@@ -2726,40 +2726,114 @@ impl World {
...
+                    #[cfg(debug_assertions)]
+                    {
+                        // if we're already panicking, log an error instead of panicking, as double-panics result in an abort
+                        #[cfg(feature = "std")]
+                        if std::thread::panicking() {
+                            log::error!("Resource `{}` was inserted during a call to World::resource_scope, which may result in unexpected behavior.\n\
+                                   In release builds, the value inserted will be overwritten at the end of the scope.",
+                                   DebugName::type_name::<R>());
+                            // return early to maintain consistent behavior with non-panicking calls in debug builds
+                            return;
```
Comment (hymm) [inline crates/bevy_ecs/src/world/mod.rs:2772]: What happens in this code branch when using `resource_scope`? I think you'd end up double panicking since `f` will panic and then the unwrap on the option will panic`. I think we need to change how resource_scope is handling the error state too.

File: crates/bevy_ecs/src/world/mod.rs [hunk 3]
```diff
@@ -2726,40 +2726,114 @@ impl World {
...
+                },
+            };
+
+            f(guard.world, value_mut)
+
+            // guard's drop impl runs here
+        };
 
-        Some(result)
+        was_successful.then_some(result)
```
Comment (hymm) [inline crates/bevy_ecs/src/world/mod.rs:2836]: This feels weird to me. Why return `None` here as we already have the `result`? At the very least the documentation should be updated to say you can get `None` if `f` panicks. But given my comment above about the double panic in `resource_scope` I'm starting to believe that `try_resource_scope`  might need to be changed to return a result instead to differentiate between when the resource doesn't exist vs when `f` has panicked.

**Comment by hymm** [comment]
Generally onboard with this change. Scoping this to `debug_assertions` means that any perf concerns are probably not worth worrying about. I do have a concern about how this changes the behavior of `resource_scope` though.

File: crates/bevy_ecs/src/world/mod.rs (hunk 3 above)
Comment (joseph-gio) [reply crates/bevy_ecs/src/world/mod.rs:2836]: Yeah the behavior is a tad strange, but it's consistent with how it works on main if the resources get cleared. I'd be down to change that in a separate PR, but my intention here was to not change any behavior for the non-panicking case

File: crates/bevy_ecs/src/world/mod.rs (hunk 2 above)
Comment (joseph-gio) [reply crates/bevy_ecs/src/world/mod.rs:2772]: My understanding is if the closure panics then the `.unwrap()` method will never get called -- it'll unwind before getting that far.

File: crates/bevy_ecs/src/world/mod.rs (hunk 3 above)
Comment (joseph-gio) [reply crates/bevy_ecs/src/world/mod.rs:2836]: > At the very least the documentation should be updated to say you can get None if f panicks

This isn't quite accurate -- if `f` panics you'll get `!`, rather than `None`. The function never completes or returns a value -- it just attempts to put the resource back into the world as it unwinds

File: crates/bevy_ecs/src/world/mod.rs
Comment (joseph-gio) [reply crates/bevy_ecs/src/world/mod.rs:2711]: I could be wrong but I don't believe that this `.unwrap_or_else(|| ())` is impacted by the changes in this PR. If `f` panics then this `unwrap()` will never get called, as the stack will unwind from the `.try_resource_scope()` before reaching the unwrap

Comment (ickshonpe) [reply crates/bevy_ecs/src/world/mod.rs:2711]: This is what happened in the naive example I tried.

**Review by ickshonpe — APPROVED** [review]
This makes sense to me and I couldn't find any problems.

File: crates/bevy_ecs/src/world/mod.rs
Comment (hymm) [reply crates/bevy_ecs/src/world/mod.rs:2711]: you're right.

File: crates/bevy_ecs/src/world/mod.rs (hunk 2 above)
Comment (hymm) [reply crates/bevy_ecs/src/world/mod.rs:2772]: Yeah you're right. I was trying to figure out when you could observe a None value from a panic and forgot the drop doesn't really return to normal control flow here.

Is there a circumstance where you could see the `None` value from when `was_successful = false`?

Comment (joseph-gio) [reply crates/bevy_ecs/src/world/mod.rs:2772]: If you call `world.clear_resources()` within the scope, then the call to `self.world.storages.resources.get_mut(self.component_id)` will return `None`, resulting in `None` being returned from the function. Bit of an odd behavior but it's how it works on main. This behavior is what the `resource_scope_resources_cleared` test I added checks for

**Review by alice-i-cecile — APPROVED** [review]
I think this is an improvement, and I'm happy to merge this for now.

**Review by hymm — APPROVED** [review]
